File: leetcode/insert-delete-getrandom-o1/solution.cpp
class RandomizedSet {
public:
    vector<int> vals;
    unordered_map<int, int> valToIdx;

    RandomizedSet() {
    }

    bool insert(int val) {
        if (valToIdx.contains(val)) {
            return false;
        }

        valToIdx[val] = vals.size();
        vals.push_back(val);
        return true;
    }

    bool remove(int val) {
        if (!valToIdx.contains(val)) {
            return false;
        }

        int pos = valToIdx[val];
        vals[pos] = vals.back();
        valToIdx[vals[pos]] = pos;
        valToIdx.erase(val);
        vals.pop_back();
        return true;
    }

    int getRandom() {
        return vals[rand() % vals.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: leetcode/insert-delete-getrandom-o1/solution.py
import random

class RandomizedSet:

    def __init__(self):
        self.idx = {}   # val -> index in nums
        self.nums = []

    def insert(self, val: int) -> bool:
        if val in self.idx:
            return False
        self.idx[val] = len(self.nums)
        self.nums.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.idx:
            return False
        i = self.idx[val]
        last = self.nums[-1]
        self.nums[i] = last
        self.idx[last] = i
        self.nums.pop()
        del self.idx[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self.nums)
